### PR TITLE
apiextensions-apiserver: simplify validation codepath

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR makes some of our validation of custom resources a bit more straightforward:

1. Moves check and error return for unstructured earlier in code (but semantically noop)
2. Simplifies control flow due to moved unstructured
3. Removes extra logic for considering multiple structural schema versions in one strategy (each stragey only handles a single version, so redundant)

Before handling a request with the strategy, teh custom resource handler always finds the storage+strategy for the specific version e.g.:
```go
	storage := crdInfo.storages[requestInfo.APIVersion].CustomResource
```

and its restOptionsGetter performs conversion

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go [hunk 1]
```diff
@@ -163,25 +164,28 @@ func copyNonMetadata(original map[string]interface{}) map[string]interface{} {
...
+	if !ok {
+		return field.ErrorList{field.Invalid(field.NewPath(""), u, fmt.Sprintf("has type %T. Must be a pointer to an Unstructured type", u))}
+	}
+
 	var errs field.ErrorList
-	errs = append(errs, a.validator.Validate(ctx, obj, a.scale)...)
+	errs = append(errs, a.validator.Validate(ctx, u, a.scale)...)
 
 	// validate embedded resources
-	if u, ok := obj.(*unstructured.Unstructured); ok {
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:170]: note to reviewers: just removed if statement and unindented

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go [hunk 2]
```diff
@@ -163,25 +164,28 @@ func copyNonMetadata(original map[string]interface{}) map[string]interface{} {
 
 // Validate validates a new CustomResource.
 func (a customResourceStrategy) Validate(ctx context.Context, obj runtime.Object) field.ErrorList {
+	u, ok := obj.(*unstructured.Unstructured)
+	if !ok {
+		return field.ErrorList{field.Invalid(field.NewPath(""), u, fmt.Sprintf("has type %T. Must be a pointer to an Unstructured type", u))}
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go:165]: note to reviewers: this was moved from early returns in `a.validator.Validate`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go [hunk 3]
```diff
@@ -44,58 +44,32 @@ type customResourceValidator struct {
...
-	if !ok {
-		return field.ErrorList{field.Invalid(field.NewPath(""), u, fmt.Sprintf("has type %T. Must be a pointer to an Unstructured type", u))}
-	}
-	accessor, err := meta.Accessor(obj)
-	if err != nil {
-		return field.ErrorList{field.Invalid(field.NewPath("metadata"), nil, err.Error())}
-	}
-
-	if errs := a.ValidateTypeMeta(ctx, u); len(errs) > 0 {
+func (a customResourceValidator) Validate(ctx context.Context, obj *unstructured.Unstructured, scale *apiextensions.CustomResourceSubresourceScale) field.ErrorList {
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go:46]: note to reviewers: signature changed to `*unstructured.Unstructured`. error returns moved to strategy

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go [hunk 4]
```diff
@@ -84,36 +85,31 @@ func (a statusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.O
...
-
 	uNew, ok := obj.(*unstructured.Unstructured)
 	if !ok {
-		return errs
+		return field.ErrorList{field.Invalid(field.NewPath(""), obj, fmt.Sprintf("has type %T. Must be a pointer to an Unstructured type", obj))}
 	}
 	uOld, ok := old.(*unstructured.Unstructured)
-	var oldObject map[string]interface{}
 	if !ok {
-		oldObject = nil
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go:97]: This is a functional change, passing with non-unstructured objects before, now it is an error. Wondering whether we anything in mind here. Scale subresource comes to mind. Is that the same strategy? Probably not.

Comment (alexzielenski) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go:97]: This was added relatively recently and seemingly as a defensive measure: https://github.com/kubernetes/kubernetes/pull/108073/files#diff-0508ef6d93f19e18a1b44e6e02ae3ff4cac9cc8065ff5c41e2bea4b9eeb6ee0bR95

Gladly would add logic to allow untypted `nil`, but curious if we actually want to?

Comment (alexzielenski) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go:97]: Also note that this error return was moved from `ValidateStatusUpdate`, so passing untyped nil was always an error.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go:97]: Right. In my own words: the `customResourceValidator` always error'ed on non-unstructured. So everywhere (and that is everywhere in this PR) where we called the `customResourceValidator` and then did conditional logic on unstructured'ness, we can return a hard error directly without regressing.

**Comment by sttts** [comment]
I think the PR is safe (and a nice simplification). The assumption is that a v1 storage can never see an object of v2. I believe that's true.

We don't check that, do we? Maybe just to be sure and not validate against a wrong schema, we could add a check for the GVR and hard-fail if it does not match?

**Comment by alexzielenski** [comment]
> Maybe just to be sure and not validate against a wrong schema, we could add a check for the GVR and hard-fail if it does not match?

https://github.com/kubernetes/kubernetes/blob/13b52a18483ff57e7e352f8bae4ef289d2f3f699/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#L154-L169

There seems to be a check already in `customResourceValidator.ValidateTypeMeta` that verifies `kind` and `apiVersion` match that of the storage.

**Comment by pacoxu** [comment]
/release-note-none
If I understand correctly, this more like a refactor/cleanup and may not need a release note.
- Add one if release note is needed.

**Pull request merged**